src/util.rs

```diff
@@ -136,88 +136,120 @@ pub fn respace(s: &str) -> String {
     s.split_whitespace().collect::<Vec<_>>().join(" ")
 }
 
-pub struct ExpandedRegister<'a> {
-    pub info: &'a RegisterInfo,
-    pub name: String,
-    pub offset: u32,
-    pub ty: Either<String, Rc<String>>,
+pub fn register_to_rust(register: &Register) -> Tokens {
+    match *register {
+        Register::Single(ref _info) => {
+            let comment = &format!(
+                "0x{:02x} - {}",
+                register.address_offset,
+                respace(&register.description)
+            )
+                [..];
+
+            let array_index_match = Regex::new(r"%s\((?P<i>[ a-zA-Z0-9]*)\)").unwrap();
+            //println!("{}", register.name.as_str());
+            let rty = Ident::new(array_index_match.replace_all(register.name.as_str(), "").to_sanitized_upper_case());
+            let reg_name = Ident::new(array_index_match.replace_all(register.name.as_str(), "$i").to_sanitized_snake_case());
+
+            quote! {
+                #[doc = #comment]
+                pub #reg_name : #rty,
+            }
+        },
+        Register::Array(ref info, ref array_info) => {
+            let has_brackets = info.name.contains("[%s]");
+            
+            let comment = &format!(
+                "0x{:02x} - {} [{}]",
+                register.address_offset,
+                respace(&register.description),
+                array_info.dim,
+            )
+                [..];
+
+            let rty = Ident::from({
+                let name = if has_brackets {
+                    info.name.replace("[%s]", "")
+                } else {
+                    info.name.replace("%s", "")
+                };
+                name.to_sanitized_upper_case().into_owned()
+            });
+
+            let reg_name = Ident::from({
+                let name = if has_brackets {
+                    info.name.replace("[%s]", "")
+                } else {
+                    info.name.replace("%s", "")
+                };
+                name.to_sanitized_snake_case().into_owned()
+            });
+
+            let length = Ident::from(array_info.dim.to_string());
+
+            quote! {
+                #[doc = #comment]
+                pub #reg_name : [#rty; #length],
+            }
+        },
+    }
 }
 
+
 /// Takes a list of "registers", some of which may actually be register arrays,
-/// and turns it into a new *sorted* (by address offset) list of registers where
-/// the register arrays have been expanded.
-pub fn expand(registers: &[Register]) -> Vec<ExpandedRegister> {
+/// and turn in into alist of registers where the register arrays have been expanded.
+pub fn expand(register: &Register) -> Vec<Register> {
```

Comment:
Given that this will now only be used with `Register::Array` it may be better not to pass `&Register` but rather the contens of the `Register::Array` variant